amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/fanout/FanOutConsumerRegistration.java

```diff
@@ -0,0 +1,242 @@
+import software.amazon.awssdk.services.kinesis.model.RegisterStreamConsumerRequest;
+import software.amazon.awssdk.services.kinesis.model.RegisterStreamConsumerResponse;
+import software.amazon.awssdk.services.kinesis.model.ResourceInUseException;
+import software.amazon.awssdk.services.kinesis.model.ResourceNotFoundException;
+import software.amazon.kinesis.annotations.KinesisClientInternalApi;
+import software.amazon.kinesis.common.KinesisRequestsBuilder;
+import software.amazon.kinesis.leases.exceptions.DependencyException;
+import software.amazon.kinesis.retrieval.AWSExceptionManager;
+import software.amazon.kinesis.retrieval.ConsumerRegistration;
+
+/**
+ *
+ */
+@RequiredArgsConstructor
+@Slf4j
+@Accessors(fluent = true)
+@KinesisClientInternalApi
+public class FanOutConsumerRegistration implements ConsumerRegistration {
+    @NonNull
+    private final KinesisAsyncClient kinesisClient;
+    private final String streamName;
+    @NonNull
+    private final String streamConsumerName;
+    private final int maxDescribeStreamSummaryRetries;
+    private final int maxDescribeStreamConsumerRetries;
+    private final int registerStreamConsumerRetries;
+    private final long retryBackoffMillis;
+
+    private String streamArn;
+    @Setter(AccessLevel.PRIVATE)
+    private String streamConsumerArn;
+
+    /**
+     * @inheritDoc
+     */
+    @Override
+    public String getOrCreateStreamConsumerArn() throws DependencyException {
+        if (StringUtils.isEmpty(streamConsumerArn)) {
+            DescribeStreamConsumerResponse response = null;
+
+            // 1. Check if consumer exists
+            try {
+                response = describeStreamConsumer();
+            } catch (ResourceNotFoundException e) {
+                log.info("StreamConsumer not found, need to create it.");
+            }
+
+            // 2. If not, register consumer
+            if (response == null) {
+                LimitExceededException finalException = null;
+                int retries = registerStreamConsumerRetries;
+                try {
+                    while (retries > 0) {
+                        finalException = null;
+                        try {
+                            final RegisterStreamConsumerResponse registerResponse = registerStreamConsumer();
+                            streamConsumerArn(registerResponse.consumer().consumerARN());
+                            break;
+                        } catch (LimitExceededException e) {
+                            // TODO: Figure out internal service exceptions
+                            log.debug("RegisterStreamConsumer call got throttled will retry.");
+                            finalException = e;
+                        }
+                        retries--;
+                    }
+
+                    // All calls got throttled, returning.
+                    if (finalException != null) {
+                        throw new DependencyException(finalException);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer

Exception [DependencyException] is explicitly thrown on this line but is caught in the outer catch clause with generic exception on line [105].  Consider adding a catch clause of this specific exception in the outer try statement.